File: send_to_tuner.ts
/**
 * Send your VAPI calls to Tuner
 * ==============================
 * Drop this file next to your VAPI server and call `sendCallToTuner(message)`
 * from inside your `end-of-call-report` handler. Every finished call — including
 * its tool calls and tool results — will show up under your Tuner agent.
 *
 * Setup (4 values, all from the Tuner dashboard):
 *   1. TUNER_BASE_URL   -> your Tuner API base URL
 *   2. TUNER_API_KEY    -> Workspace Settings > API Keys           (starts with "tr_api_")
 *   3. TUNER_WORKSPACE  -> Workspace > General Settings
 *   4. TUNER_AGENT_ID   -> create a "Custom API" agent in Tuner, then
 *                          Agent Settings > Agent Connection > Agent ID
 *
 * Requirements: Node 18+ (uses the built-in fetch). No npm dependencies.
 */

// =============================================================================
// CONFIG — fill these in
// =============================================================================
const TUNER_BASE_URL = "https://api.usetuner.ai"; // your Tuner API base URL
const TUNER_API_KEY = "tr_api_f479e62c-0798-4515-993d-556b46bc8b64";
const TUNER_WORKSPACE = 1420; // your workspace id (number)
const TUNER_AGENT_ID = "fa4da74c-f775-4d20-897f-887b7274b067";


type Json = Record<string, any>;

// =============================================================================
// THE FUNCTION — call this from your end-of-call handler
// =============================================================================
/**
 * Forward one VAPI `end-of-call-report` message to Tuner.
 *
 * `message` is the `body.message` object VAPI posts to your webhook.
 * Safe to call inline — it never throws; failures are logged, not thrown.
 */
export async function sendCallToTuner(message: Json): Promise<void> {
  try {
    const call: Json = message.call ?? {};
    const artifact: Json = message.artifact ?? {};

    const callId = call.id;
    if (!callId) {
      console.warn("Tuner: end-of-call-report has no call id — call will not be sent");
      return;
    }

    // VAPI keeps the timed, structured transcript under artifact.messages.
    const rawMessages: Json[] = artifact.messages ?? message.messages ?? [];
    const callStartMs = toEpochMs(message.startedAt);
    let segments = buildSegments(rawMessages, callStartMs);
    if (segments.length === 0) {
      console.warn(`Tuner: no transcript segments found — call ${callId} will not be sent`);
      return;
    }

    // Attach per-turn latency: STT to the customer turn, LLM/TTS/e2e to the agent reply.
    const turnLatencies = artifact.performanceMetrics?.turnLatencies;
    if (turnLatencies?.length) {
      enrichWithTurnLatencies(segments, turnLatencies);
    }
    // Flag agent turns the customer talked over (barge-in).
    markInterruptions(segments);
    segments = segments.map(stripNull);

    const stereoUrl = message.stereoRecordingUrl ?? artifact.stereoRecordingUrl ?? null;
    const recordingUrl = message.recordingUrl ?? artifact.recordingUrl ?? stereoUrl;
    if (!recordingUrl) {
      console.warn(`Tuner: no recording URL — call ${callId} will not be sent to Tuner`);
      return;
    }

    const { callAnalysis, callSuccessful } = mapAnalysis(message);
    const payload: Json = {
      call_id: callId,
      call_type: mapCallType(call.type),
      start_timestamp: callStartMs,
      end_timestamp: toEpochMs(message.endedAt),
      recording_url: recordingUrl,
      recording_multi_channel_url: stereoUrl,
      transcript: message.transcript ?? artifact.transcript ?? null,
      transcript_with_tool_calls: segments,
      call_status: "call_ended",
      disconnection_reason: (message.endedReason ?? "").slice(0, 100) || null,
      caller_phone_number: call.customer?.number ?? null,
      call_successful: callSuccessful,
      call_analysis: callAnalysis,
      call_cost: costInCents(message.cost),
    };
    // Drop empty optional fields so we only send what we actually have.
    for (const key of Object.keys(payload)) {
      if (payload[key] === null || payload[key] === undefined) delete payload[key];
    }

    const params = new URLSearchParams({
      workspace_id: String(TUNER_WORKSPACE),
      agent_remote_identifier: TUNER_AGENT_ID,
    });
    const response = await fetch(`${TUNER_BASE_URL}/api/v1/public/call?${params}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${TUNER_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(15_000),
    });

    if (response.status === 200 || response.status === 201) {
      console.info(`Tuner: sent call ${callId}`);
    } else if (response.status === 409) {
      console.info(`Tuner: call ${callId} already sent`); // idempotent, fine
    } else {
      console.error(`Tuner: failed (${response.status}) ${await response.text()}`);
    }
  } catch (err) {
    // never let Tuner break your call handling
    console.error("Tuner: error sending call:", err);
  }
}

// =============================================================================
// CALL-LEVEL MAPPING — mirrors how Tuner maps VAPI calls natively.
// =============================================================================
// VAPI call types -> Tuner's canonical snake_case values; unknown types pass through.
const VAPI_CALL_TYPE_MAP: Record<string, string> = {
  webCall: "web_call",
  phoneCall: "phone_call",
  sipCall: "sip_call",
};

/** Normalize VAPI's camelCase call type to Tuner's snake_case. */
function mapCallType(vapiType: string | undefined | null): string {
  if (!vapiType) return "phone_call";
  return VAPI_CALL_TYPE_MAP[vapiType] ?? vapiType;
}

/** Build Tuner's call_analysis ({summary, success_evaluation}) and call_successful. */
function mapAnalysis(message: Json): { callAnalysis: Json | null; callSuccessful: boolean | null } {
  const callAnalysis: Json = {};
  let callSuccessful: boolean | null = null;

  const analysis: Json = message.analysis ?? {};
  if (analysis.summary) {
    callAnalysis.summary = analysis.summary;
  }
  const successEval = analysis.successEvaluation;
  if (successEval !== undefined && successEval !== null) {
    callAnalysis.success_evaluation = successEval;
    callSuccessful = String(successEval).toLowerCase() === "true";
  }

  // Fall back to the top-level summary if analysis didn't carry one.
  if (message.summary && !("summary" in callAnalysis)) {
    callAnalysis.summary = message.summary;
  }

  return {
    callAnalysis: Object.keys(callAnalysis).length ? callAnalysis : null,
    callSuccessful,
  };
}

/** VAPI reports cost in dollars; Tuner expects cents. */
function costInCents(cost: unknown): number | null {
  if (typeof cost !== "number" || Number.isNaN(cost)) return null;
  return Math.round(cost * 100 * 100) / 100; // round to 2 decimals
}

/** Drop null/undefined values recursively so segments only carry fields that exist. */
function stripNull<T>(obj: T): T {
  if (Array.isArray(obj)) {
    return obj.map(stripNull) as T;
  }
  if (obj !== null && typeof obj === "object") {
    const out: Json = {};
    for (const [k, v] of Object.entries(obj)) {
      if (v !== null && v !== undefined) out[k] = stripNull(v);
    }
    return out as T;
  }
  return obj;
}

// =============================================================================
// MAPPING — VAPI messages -> Tuner's transcript timeline.
// Mirrors Tuner's internal VAPI mapper so user/agent speech, tool calls, and
// tool results are all captured. VAPI roles map to Tuner roles like so:
//   user  -> user        bot         -> agent
//   tool_calls -> agent_function     tool_call_result -> agent_result
//   system -> dropped
// =============================================================================
function buildSegments(rawMessages: Json[], callStartMs: number | null): Json[] {
  const segments: Json[] = [];
  for (const m of rawMessages) {
    if (typeof m !== "object" || m === null) continue;
    const role = m.role;
    if (role === "user") {
      const seg = speechSegment(m, "user", callStartMs);
      if (seg) segments.push(seg);
    } else if (role === "bot" || role === "assistant") {
      const seg = speechSegment(m, "agent", callStartMs);
      if (seg) segments.push(seg);
    } else if (role === "tool_calls") {
      segments.push(...toolCallSegments(m));
    } else if (role === "tool_call_result") {
      segments.push(toolResultSegment(m));
    }
    // "system" and anything else is intentionally skipped.
  }
  return segments;
}

/**
 * A user or agent spoken turn, with timing (and word-level timing if present).
 *
 * Tuner requires timing on every spoken turn. A segment with no resolvable timing
 * would reject the whole call, so we skip it instead — same as Tuner does natively.
 */
function speechSegment(m: Json, role: string, callStartMs: number | null): Json | null {
  const { startMs, endMs, durationMs } = segmentTiming(m, callStartMs);
  const segWords = words(m.metadata, callStartMs);

  const hasTiming = startMs !== null && (endMs !== null || durationMs !== null);
  if (!hasTiming && !segWords) {
    console.warn(`Tuner: skipping untimed ${role} segment: ${String(m.message ?? "").slice(0, 50)}`);
    return null;
  }

  const seg: Json = { role, text: m.message };
  if (startMs !== null) seg.start_ms = startMs;
  if (endMs !== null) seg.end_ms = endMs;
  if (durationMs !== null) seg.duration_ms = durationMs;
  if (segWords) seg.words = segWords;
  return seg;
}

/** One agent_function segment per tool the agent invoked. */
function toolCallSegments(m: Json): Json[] {
  const seconds = m.secondsFromStart;
  const startMs = seconds !== undefined && seconds !== null ? Math.ceil(seconds * 1000) : null;
  const out: Json[] = [];
  for (const tc of m.toolCalls ?? []) {
    const fn: Json = tc.function ?? {};
    const params = maybeJson(fn.arguments);
    out.push({
      role: "agent_function",
      start_ms: startMs,
      end_ms: startMs,
      tool: {
        name: fn.name,
        request_id: tc.id,
        params: typeof params === "object" && params !== null && !Array.isArray(params) ? params : null,
        start_ms: startMs,
        end_ms: startMs,
      },
    });
  }
  return out;
}

/** The result returned by a tool. Non-object results are wrapped so nothing is lost. */
function toolResultSegment(m: Json): Json {
  const seconds = m.secondsFromStart;
  const startMs = seconds !== undefined && seconds !== null ? Math.ceil(seconds * 1000) : null;
  const parsed = maybeJson(m.result);

  let result: Json | null;
  if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
    result = parsed;
  } else if ((Array.isArray(parsed) && parsed.length) || (typeof parsed === "string" && parsed)) {
    result = { value: parsed };
  } else {
    result = null;
  }

  const { isError, errorMsg } = detectToolError(m, parsed);
  const tool: Json = {
    name: m.name,
    request_id: m.toolCallId,
    result,
    is_error: isError,
    start_ms: startMs,
    end_ms: startMs,
  };
  if (errorMsg) tool.error = errorMsg;
  return { role: "agent_result", start_ms: startMs, end_ms: startMs, tool };
}

/**
 * Detect whether a tool result represents an error, and extract its message.
 *
 * VAPI has no single standard error flag, so we check the explicit signals that
 * show up in practice — an error field on the message itself, or an error-shaped
 * result payload — rather than guessing from free text.
 */
function detectToolError(m: Json, parsed: unknown): { isError: boolean; errorMsg: string | null } {
  // 1. Explicit flags on the VAPI message (extra fields are allowed by VAPI).
  const msgError = m.error;
  if (typeof msgError === "string" && msgError) {
    return { isError: true, errorMsg: msgError };
  }
  if (m.error === true || m.isError === true || m.is_error === true) {
    return { isError: true, errorMsg: typeof msgError === "string" ? msgError : null };
  }
  if (m.success === false) {
    return { isError: true, errorMsg: null };
  }

  // 2. Error-shaped result payload, e.g. {"error": "..."} or {"status": "failed"}.
  if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
    const p = parsed as Json;
    if (p.error) {
      return { isError: true, errorMsg: typeof p.error === "string" ? p.error : null };
    }
    if (p.success === false) {
      return { isError: true, errorMsg: null };
    }
    const status = p.status;
    if (typeof status === "string" && ["error", "failed", "failure"].includes(status.toLowerCase())) {
      return { isError: true, errorMsg: typeof p.message === "string" ? p.message : null };
    }
  }

  return { isError: false, errorMsg: null };
}

/**
 * Attach VAPI's turnLatencies to the right rows.
 *
 * VAPI emits ONE turnLatencies entry per completed agent response turn, and a
 * turn can span several messages. The grouping below was verified against
 * VAPI's internal call logs (every log event carries a turnId):
 *  - The canned firstMessage greeting gets NO entry.
 *  - A turn = one run of agent activity triggered by customer speech.
 *  - A tool call that follows agent speech in the same run starts a NEW turn
 *    (it is a fresh model decision). A tool call with no agent speech before
 *    it belongs to the current turn — its filler speech ("Give me a moment"),
 *    the tool result, and the post-result reply all stay in that same turn.
 *  - On tool-call turns VAPI reports modelLatency/voiceLatency as 0.
 *
 * Per entry: stt -> the customer row that triggered the turn (first turn wins
 * when one customer row triggered two turns); llm/tts/e2e -> the turn's first
 * agent speech row. Continuation rows stay badge-less. Zeros are kept.
 */
function enrichWithTurnLatencies(segments: Json[], turnLatencies: Json[]): void {
  // --- group rows into turns ---
  type Turn = { trigger: Json | null; rows: Json[]; hasSpeech: boolean };
  const turns: Turn[] = [];
  let cur: Turn = { trigger: null, rows: [], hasSpeech: false };
  const flush = () => {
    if (cur.rows.length) turns.push(cur);
  };
  for (const seg of segments) {
    if (seg.role === "user") {
      flush();
      cur = { trigger: seg, rows: [], hasSpeech: false };
    } else if (seg.role === "agent") {
      cur.rows.push(seg);
      cur.hasSpeech = true;
    } else if (seg.role === "agent_function" && cur.hasSpeech) {
      flush();
      cur = { trigger: cur.trigger, rows: [seg], hasSpeech: false };
    } else {
      cur.rows.push(seg); // agent_result, or a tool call opening the turn
    }
  }
  flush();

  // Agent speech before any customer speech (the greeting) never gets an entry.
  const eligible = turns.filter((t) => t.trigger !== null);

  const count = Math.min(eligible.length, turnLatencies.length);
  for (let i = 0; i < count; i++) {
    const turn = eligible[i];
    const tl = turnLatencies[i];

    const trigger = turn.trigger!;
    if (tl.transcriberLatency != null && trigger.metadata?.stt_node_ttfb === undefined) {
      trigger.metadata = { ...(trigger.metadata ?? {}), stt_node_ttfb: tl.transcriberLatency };
    }

    const speech = turn.rows.find((r) => r.role === "agent");
    if (!speech) continue; // e.g. an endCall turn with no spoken reply
    const metadata: Json = {};
    if (tl.turnLatency != null) metadata.e2e_latency = tl.turnLatency;
    if (tl.modelLatency != null) metadata.llm_node_ttft = tl.modelLatency;
    if (tl.voiceLatency != null) metadata.tts_node_ttfb = tl.voiceLatency;
    if (Object.keys(metadata).length) speech.metadata = { ...(speech.metadata ?? {}), ...metadata };
  }
}

// A customer turn must start at least this many ms before the agent's turn ends
// to count as a barge-in (filters out word-boundary noise and short backchannels).
const INTERRUPTION_THRESHOLD_MS = 200;

/**
 * Flag agent turns the customer talked over.
 *
 * A barge-in is the customer starting to speak before the agent finished. We
 * compare each customer turn against the agent turn right before it; if it
 * starts before that agent turn ended (beyond a small threshold), the agent was
 * interrupted. The flag is stamped on the AGENT segment that got cut off.
 */
function markInterruptions(segments: Json[]): void {
  let lastAgentSeg: Json | null = null;
  for (const seg of segments) {
    if (seg.role === "agent") {
      lastAgentSeg = seg;
    } else if (seg.role === "user") {
      if (
        lastAgentSeg &&
        lastAgentSeg.end_ms != null &&
        seg.start_ms != null &&
        seg.start_ms < lastAgentSeg.end_ms - INTERRUPTION_THRESHOLD_MS
      ) {
        lastAgentSeg.metadata = { ...(lastAgentSeg.metadata ?? {}), interrupted: true };
      }
      lastAgentSeg = null; // each customer turn compares against a fresh agent turn
    }
    // tool segments: ignore, keep the current agent turn as the candidate
  }
}

// =============================================================================
// TIMING HELPERS
// =============================================================================
/**
 * Resolve (start_ms, end_ms, duration_ms) relative to call start.
 *
 * Prefer secondsFromStart (already relative to call start). Fall back to the
 * absolute epoch `time`/`endTime` fields when call start is known.
 */
function segmentTiming(
  m: Json,
  callStartMs: number | null,
): { startMs: number | null; endMs: number | null; durationMs: number | null } {
  let startMs: number | null = null;
  let endMs: number | null = null;
  let durationMs: number | null = null;

  const seconds = m.secondsFromStart;
  if (seconds !== undefined && seconds !== null) {
    startMs = Math.ceil(seconds * 1000);
  }
  if (m.duration !== undefined && m.duration !== null) {
    durationMs = Math.ceil(m.duration);
  }

  if (startMs !== null && durationMs !== null) {
    endMs = startMs + durationMs;
  } else if (callStartMs !== null) {
    if (m.time !== undefined && m.time !== null && startMs === null) {
      startMs = Math.ceil(m.time - callStartMs);
    }
    if (m.endTime !== undefined && m.endTime !== null) {
      endMs = Math.ceil(m.endTime - callStartMs);
    }
    if (startMs !== null && endMs !== null) {
      durationMs = endMs - startMs;
    }
  }
  return { startMs, endMs, durationMs };
}

/**
 * Word-level timing from VAPI's wordLevelConfidence.
 *
 * VAPI has shipped word start/end in two formats: seconds from call start
 * (current payloads, e.g. 5.98) and epoch milliseconds (older payloads,
 * e.g. 1774621268400). Detect the unit by magnitude so neither produces
 * garbage timestamps.
 */
function words(metadata: unknown, callStartMs: number | null): Json[] | null {
  if (typeof metadata !== "object" || metadata === null) return null;
  const raw = (metadata as Json).wordLevelConfidence;
  if (!Array.isArray(raw) || raw.length === 0) return null;

  const toRelMs = (v: number): number => {
    if (v > 1e11) return Math.ceil(v - (callStartMs ?? 0)); // epoch ms
    if (v < 1e5) return Math.ceil(v * 1000); // seconds from call start
    return Math.ceil(v); // already ms from call start
  };

  const out: Json[] = [];
  for (const w of raw) {
    if (typeof w !== "object" || w === null || !w.word || typeof w.start !== "number" || typeof w.end !== "number") {
      continue;
    }
    out.push({
      word: w.word,
      start_ms: toRelMs(w.start),
      end_ms: toRelMs(w.end),
      confidence: w.confidence ?? null,
    });
  }
  return out.length ? out : null;
}

/** VAPI sends tool arguments/results as JSON strings; parse them, else pass through. */
function maybeJson(value: unknown): unknown {
  if (typeof value !== "string") return value;
  try {
    return JSON.parse(value);
  } catch {
    return value;
  }
}

/** Accept VAPI's ISO string ('2026-06-10T...Z') or a number; return epoch ms. */
function toEpochMs(value: unknown): number | null {
  if (value === undefined || value === null) return null;
  if (typeof value === "number") return Math.trunc(value);
  const ms = Date.parse(String(value));
  return Number.isNaN(ms) ? null : ms;
}

File: send_to_tuner.py
"""
Send your VAPI calls to Tuner
==============================
Drop this file next to your VAPI server and call `send_call_to_tuner(message)`
from inside your `end-of-call-report` handler. Every finished call — including
its tool calls and tool results — will show up under your Tuner agent.

Setup (4 values, all from the Tuner dashboard):
  1. TUNER_BASE_URL   -> your Tuner API base URL
  2. TUNER_API_KEY    -> Workspace Settings > API Keys           (starts with "tr_api_")
  3. TUNER_WORKSPACE  -> Workspace > General Settings
  4. TUNER_AGENT_ID   -> create a "Custom API" agent in Tuner, then
                         Agent Settings > Agent Connection > Agent ID

Install:  pip install requests
"""

import json
import logging
import math
from datetime import datetime

import requests

logger = logging.getLogger("tuner")

# =============================================================================
# CONFIG — fill these in
# =============================================================================
TUNER_BASE_URL = "https://api.tuner.ai"  # your Tuner API base URL
TUNER_API_KEY = "tr_api_xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
TUNER_WORKSPACE = 0  # your workspace id (int)
TUNER_AGENT_ID = "your-custom-api-agent-id"


# =============================================================================
# THE FUNCTION — call this from your end-of-call handler
# =============================================================================
def send_call_to_tuner(message: dict) -> None:
    """Forward one VAPI `end-of-call-report` message to Tuner.

    `message` is the `body["message"]` object VAPI posts to your webhook.
    Safe to call inline — it never raises; failures are logged, not thrown.
    """
    try:
        call = message.get("call", {})
        artifact = message.get("artifact", {})

        call_id = call.get("id")
        if not call_id:
            logger.warning(
                "Tuner: end-of-call-report has no call id — call will not be sent"
            )
            return

        # VAPI keeps the timed, structured transcript under artifact.messages.
        raw_messages = artifact.get("messages") or message.get("messages") or []
        call_start_ms = _to_epoch_ms(message.get("startedAt"))
        segments = _build_segments(raw_messages, call_start_ms)
        if not segments:
            logger.warning(
                "Tuner: no transcript segments found — call %s will not be sent",
                call_id,
            )
            return

        # Attach per-turn latency (STT/LLM/TTS/e2e) to each agent turn.
        turn_latencies = (artifact.get("performanceMetrics") or {}).get("turnLatencies")
        if turn_latencies:
            _enrich_with_turn_latencies(segments, turn_latencies)
        segments = [_strip_none(s) for s in segments]

        stereo_url = message.get("stereoRecordingUrl") or artifact.get(
            "stereoRecordingUrl"
        )
        recording_url = (
            message.get("recordingUrl") or artifact.get("recordingUrl") or stereo_url
        )
        if not recording_url:
            logger.warning(
                "Tuner: no recording URL — call %s will not be sent to Tuner", call_id
            )
            return

        call_analysis, call_successful = _map_analysis(message)
        payload = {
            "call_id": call_id,
            "call_type": _map_call_type(call.get("type")),
            "start_timestamp": call_start_ms,
            "end_timestamp": _to_epoch_ms(message.get("endedAt")),
            "recording_url": recording_url,
            "recording_multi_channel_url": stereo_url,
            "transcript": message.get("transcript") or artifact.get("transcript"),
            "transcript_with_tool_calls": segments,
            "call_status": "call_ended",
            "disconnection_reason": (message.get("endedReason") or "")[:100] or None,
            "caller_phone_number": call.get("customer", {}).get("number"),
            "call_successful": call_successful,
            "call_analysis": call_analysis,
            "call_cost": _cost_in_cents(message.get("cost")),
        }
        # Drop empty optional fields so we only send what we actually have.
        payload = {k: v for k, v in payload.items() if v is not None}

        response = requests.post(
            f"{TUNER_BASE_URL}/api/v1/public/call",
            params={
                "workspace_id": TUNER_WORKSPACE,
                "agent_remote_identifier": TUNER_AGENT_ID,
            },
            headers={"Authorization": f"Bearer {TUNER_API_KEY}"},
            json=payload,
            timeout=15,
        )
        if response.status_code in (200, 201):
            logger.info("Tuner: sent call %s", call_id)
        elif response.status_code == 409:
            logger.info("Tuner: call %s already sent", call_id)  # idempotent, fine
        else:
            logger.error("Tuner: failed (%s) %s", response.status_code, response.text)
    except Exception as exc:  # never let Tuner break your call handling
        logger.error("Tuner: error sending call: %s", exc)


# =============================================================================
# CALL-LEVEL MAPPING — mirrors how Tuner maps VAPI calls natively.
# =============================================================================
# VAPI call types -> Tuner's canonical snake_case values; unknown types pass through.
VAPI_CALL_TYPE_MAP = {
    "webCall": "web_call",
    "phoneCall": "phone_call",
    "sipCall": "sip_call",
}


def _map_call_type(vapi_type) -> str:
    """Normalize VAPI's camelCase call type to Tuner's snake_case."""
    if not vapi_type:
        return "phone_call"
    return VAPI_CALL_TYPE_MAP.get(vapi_type, vapi_type)


def _map_analysis(message: dict) -> tuple[dict | None, bool | None]:
    """Build Tuner's call_analysis ({summary, success_evaluation}) and call_successful."""
    call_analysis: dict = {}
    call_successful = None

    analysis = message.get("analysis") or {}
    if analysis.get("summary"):
        call_analysis["summary"] = analysis["summary"]
    success_eval = analysis.get("successEvaluation")
    if success_eval is not None:
        call_analysis["success_evaluation"] = success_eval
        call_successful = str(success_eval).lower() == "true"

    # Fall back to the top-level summary if analysis didn't carry one.
    if message.get("summary") and "summary" not in call_analysis:
        call_analysis["summary"] = message["summary"]

    return call_analysis or None, call_successful


def _cost_in_cents(cost) -> float | None:
    """VAPI reports cost in dollars; Tuner expects cents."""
    if not isinstance(cost, (int, float)):
        return None
    return round(cost * 100, 2)


def _strip_none(obj):
    """Drop None values recursively so segments only carry fields that exist."""
    if isinstance(obj, dict):
        return {k: _strip_none(v) for k, v in obj.items() if v is not None}
    if isinstance(obj, list):
        return [_strip_none(x) for x in obj]
    return obj


# =============================================================================
# MAPPING — VAPI messages -> Tuner's transcript timeline.
# Mirrors Tuner's internal VAPI mapper so user/agent speech, tool calls, and
# tool results are all captured. VAPI roles map to Tuner roles like so:
#   user  -> user        bot         -> agent
#   tool_calls -> agent_function     tool_call_result -> agent_result
#   system -> dropped
# =============================================================================
def _build_segments(raw_messages: list, call_start_ms: int | None) -> list[dict]:
    """Convert VAPI's message list into Tuner transcript segments."""
    segments: list[dict] = []
    for m in raw_messages:
        if not isinstance(m, dict):
            continue
        role = m.get("role")
        if role == "user":
            seg = _speech_segment(m, "user", call_start_ms)
            if seg:
                segments.append(seg)
        elif role in ("bot", "assistant"):
            seg = _speech_segment(m, "agent", call_start_ms)
            if seg:
                segments.append(seg)
        elif role == "tool_calls":
            segments.extend(_tool_call_segments(m))
        elif role == "tool_call_result":
            segments.append(_tool_result_segment(m))
        # "system" and anything else is intentionally skipped.
    return segments


def _speech_segment(m: dict, role: str, call_start_ms: int | None) -> dict | None:
    """A user or agent spoken turn, with timing (and word-level timing if present).

    Tuner requires timing on every spoken turn. A segment with no resolvable timing
    would reject the whole call, so we skip it instead — same as Tuner does natively.
    """
    start_ms, end_ms, duration_ms = _segment_timing(m, call_start_ms)
    words = _words(m.get("metadata"), call_start_ms)

    has_timing = start_ms is not None and (
        end_ms is not None or duration_ms is not None
    )
    if not has_timing and not words:
        logger.warning(
            "Tuner: skipping untimed %s segment: %.50r", role, m.get("message") or ""
        )
        return None

    seg = {"role": role, "text": m.get("message")}
    if start_ms is not None:
        seg["start_ms"] = start_ms
    if end_ms is not None:
        seg["end_ms"] = end_ms
    if duration_ms is not None:
        seg["duration_ms"] = duration_ms
    if words:
        seg["words"] = words
    return seg


def _tool_call_segments(m: dict) -> list[dict]:
    """One agent_function segment per tool the agent invoked."""
    seconds = m.get("secondsFromStart")
    start_ms = math.ceil(seconds * 1000) if seconds is not None else None
    out = []
    for tc in m.get("toolCalls", []):
        fn = tc.get("function", {})
        params = _maybe_json(fn.get("arguments"))
        out.append(
            {
                "role": "agent_function",
                "start_ms": start_ms,
                "end_ms": start_ms,
                "tool": {
                    "name": fn.get("name"),
                    "request_id": tc.get("id"),
                    "params": params if isinstance(params, dict) else None,
                    "start_ms": start_ms,
                    "end_ms": start_ms,
                },
            }
        )
    return out


def _tool_result_segment(m: dict) -> dict:
    """The result returned by a tool. Non-dict results are wrapped so nothing is lost."""
    seconds = m.get("secondsFromStart")
    start_ms = math.ceil(seconds * 1000) if seconds is not None else None
    parsed = _maybe_json(m.get("result"))
    if isinstance(parsed, dict):
        result = parsed
    elif isinstance(parsed, (list, str)) and parsed:
        result = {"value": parsed}
    else:
        result = None

    is_error, error_msg = _detect_tool_error(m, parsed)
    tool = {
        "name": m.get("name"),
        "request_id": m.get("toolCallId"),
        "result": result,
        "is_error": is_error,
        "start_ms": start_ms,
        "end_ms": start_ms,
    }
    if error_msg:
        tool["error"] = error_msg
    return {
        "role": "agent_result",
        "start_ms": start_ms,
        "end_ms": start_ms,
        "tool": tool,
    }


def _detect_tool_error(m: dict, parsed) -> tuple[bool, str | None]:
    """Detect whether a tool result represents an error, and extract its message.

    VAPI has no single standard error flag, so we check the explicit signals that
    show up in practice — an error field on the message itself, or an error-shaped
    result payload — rather than guessing from free text.
    """
    # 1. Explicit flags on the VAPI message (extra fields are allowed by VAPI).
    msg_error = m.get("error")
    if isinstance(msg_error, str) and msg_error:
        return True, msg_error
    if any(m.get(k) is True for k in ("error", "isError", "is_error")):
        return True, msg_error if isinstance(msg_error, str) else None
    if m.get("success") is False:
        return True, None

    # 2. Error-shaped result payload, e.g. {"error": "..."} or {"status": "failed"}.
    if isinstance(parsed, dict):
        res_error = parsed.get("error")
        if res_error:
            return True, res_error if isinstance(res_error, str) else None
        if parsed.get("success") is False:
            return True, None
        status = parsed.get("status")
        if isinstance(status, str) and status.lower() in ("error", "failed", "failure"):
            msg = parsed.get("message")
            return True, msg if isinstance(msg, str) else None

    return False, None


def _enrich_with_turn_latencies(segments: list[dict], turn_latencies: list) -> None:
    """Stamp each agent turn with its latency breakdown, in transcript order.

    VAPI's performanceMetrics.turnLatencies is ordered by agent turn, so we match
    the Nth entry to the Nth agent segment.
    """
    idx = 0
    for seg in segments:
        if seg.get("role") != "agent":
            continue
        if idx >= len(turn_latencies):
            break
        tl = turn_latencies[idx]
        metadata = {
            "e2e_latency": tl.get("turnLatency"),
            "stt_node_ttfb": tl.get("transcriberLatency"),
            "llm_node_ttft": tl.get("modelLatency"),
            "tts_node_ttfb": tl.get("voiceLatency"),
        }
        metadata = {k: v for k, v in metadata.items() if v is not None}
        if metadata:
            seg["metadata"] = metadata
        idx += 1


# =============================================================================
# TIMING HELPERS
# =============================================================================
def _segment_timing(m: dict, call_start_ms: int | None):
    """Resolve (start_ms, end_ms, duration_ms) relative to call start.

    Prefer secondsFromStart (already relative to call start). Fall back to the
    absolute epoch `time`/`endTime` fields when call start is known.
    """
    start_ms = end_ms = duration_ms = None
    seconds = m.get("secondsFromStart")
    if seconds is not None:
        start_ms = math.ceil(seconds * 1000)
    if m.get("duration") is not None:
        duration_ms = math.ceil(m["duration"])

    if start_ms is not None and duration_ms is not None:
        end_ms = start_ms + duration_ms
    elif call_start_ms is not None:
        if m.get("time") is not None and start_ms is None:
            start_ms = math.ceil(m["time"] - call_start_ms)
        if m.get("endTime") is not None:
            end_ms = math.ceil(m["endTime"] - call_start_ms)
        if start_ms is not None and end_ms is not None:
            duration_ms = end_ms - start_ms
    return start_ms, end_ms, duration_ms


def _words(metadata, call_start_ms: int | None) -> list[dict] | None:
    """Word-level timing from VAPI's wordLevelConfidence (start/end are epoch ms)."""
    if not isinstance(metadata, dict):
        return None
    raw = metadata.get("wordLevelConfidence")
    if not isinstance(raw, list) or not raw:
        return None
    words = []
    for w in raw:
        if (
            not isinstance(w, dict)
            or not w.get("word")
            or "start" not in w
            or "end" not in w
        ):
            continue
        offset = call_start_ms or 0
        words.append(
            {
                "word": w.get("word"),
                "start_ms": math.ceil(w["start"] - offset),
                "end_ms": math.ceil(w["end"] - offset),
                "confidence": w.get("confidence"),
            }
        )
    return words or None


def _maybe_json(value):
    """VAPI sends tool arguments/results as JSON strings; parse them, else pass through."""
    if not isinstance(value, str):
        return value
    try:
        return json.loads(value)
    except (json.JSONDecodeError, ValueError):
        return value


def _to_epoch_ms(value) -> int | None:
    """Accept VAPI's ISO string ('2026-06-10T...Z') or a number; return epoch ms."""
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return int(value)
    return int(datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1000)
